File: leetcode/python/79. 单词搜索.py
#!/bin/bash python
"""
79. 单词搜索

思路：
也是标准的回溯法
"""

class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        标志的回溯法，flag标记是否遍历过
        """
        # 特殊情况
        if board == [] or word == "":
            return False
        
        def dfs(board, word, flag, dires, row, col, i, j, index):
            """
            回溯法
            返回是否找到了
            """
            # 如果index已经超出了，则已经找到
            if index >= len(word):
                return True
            # 如果越界，或者已经遍历过，或者没有匹配上，则没有找到
            if i < 0 or j < 0 or i >= row or j >= col or flag[i][j] == True or board[i][j] != word[index]:
                return False
            # 没有越界，没有遍历过，已经匹配上了
            flag[i][j] = True
            for d in dires:
                x, y = i + d[0], j + d[1]
                if dfs(board, word, flag, dires, row, col, x, y, index + 1):
                    return True
            flag[i][j] = False
            return False
        
        # 初始化
        row = len(board)
        col = len(board[0])
        flag = [[False] * col for _ in range(row)]
        dires = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        # 找到入口开始遍历
        for i in range(row):
            for j in range(col):
                if dfs(board, word, flag, dires, row, col, i, j, 0):
                    return True
        return False

File: leetcode/cpp/79. 单词搜索.cpp
class Solution {
public:
    bool dfs(const vector<vector<char>>& board, const string& word, const vector<vector<int>>& dires, vector<vector<bool>>& flag, int row, int col, int i, int j, int index) {
        // 如果index越界则说明找到了
        if (index >= word.size()) {
            return true;
        }

        // 如果越界，已经遍历，没有匹配上则没有找到
        if (i < 0 || j < 0 || i >= row || j >= col || flag[i][j] == true || board[i][j] != word[index]) {
            return false;
        }

        // 如果没有越界，没有遍历过，已经匹配，则回溯下一层
        flag[i][j] = true;
        for (const auto & d : dires) {
            int x = i + d[0];
            int y = j + d[1];
            if (dfs(board, word, dires, flag, row, col, x, y, index + 1)) {
                return true;
            }
        }
        flag[i][j] = false;
        return false;
    }

    bool exist(vector<vector<char>>& board, string word) {
        // 特殊情况
        if (board.empty() || word.empty()) {
            return false;
        }

        // 初始化
        int row = board.size();
        int col = board[0].size();
        vector<vector<bool>> flag(row, vector<bool>(col, false));
        vector<vector<int>> dires = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        // 找到入口并开始回溯
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (dfs(board, word, dires, flag, row, col, i, j, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
};